sql/migrations/1_initial.sql

```diff
@@ -0,0 +1,27 @@
+CREATE TABLE blocks ( 
+	id CHAR(20) PRIMARY KEY,
+	layer INT,
+	validity SMALL INT,
+	block    BLOB
+);
+
+CREATE INDEX blocks_by_layer ON blocks(layer);
+CREATE INDEX blocks_by_validity ON blocks(layer, validity) where validity = 1;
+
+CREATE TABLE ballots ( 
+	id CHAR(20) PRIMARY KEY,
+	layer INT,
+    signature VARCHAR,
+    pubkey VARCHAR,
+	ballot BLOB
+);
+
+CREATE INDEX ballots_by_layer ON ballots(layer);
+
+CREATE TABLE layers (
+    id INT PRIMARY KEY,
+    hare_output VARCHAR,
+    status SMALL INT
```

Comment:
what you described is covered using a specific API, i don't understand why it is natural to create more tables if same functionality can be provided with less bloat

Reply:
i don't think the number of tables is that important. i don't understand why layer should have a status attached to it.
they are not attribute of the layer but attributes of the mesh. a given layer can be `latest` and `processed` at the same time.

Reply:
> i don't think the number of tables is that important

i think status reflects a stage in the life-cycle of the layer, latest -> processed -> applied. it always moves from left to right, so if the layer is processed layer it also implies that it can be latest.

> they are not attribute of the layer but attributes of the mesh.

i don't see why the are not attribute of the layer. for instance only applied layers has a hash and aggregated hash. processed layer has a hare_output. we actually could determine status just from looking at fields, but that would be more complex

Reply:
so if layer 100 is set to `latest` first, and then set to `processed` later. when the app asked for latest layer, do you return 99 or 100?

Reply:
> i think status reflects a stage in the life-cycle of the layer, latest -> processed -> applied. it always moves from left to right, so if the layer is processed layer it also implies that it can be latest.

except you don't guarantee that in code. and there is no where in the app we asked "what is the status of this layer?"
in our app. that's why this seem unnatural to me.

the only query in the app is about the mesh itself. what is the mesh's processed/applied/latest layer. for the purpose of spacemesh app, i don't think it's a property of the layer. it is the property of the mesh.

Reply:
what does it mean when you say the layer is "latest"? not every layer get set to the latest. 
for example, if there is unstable network, we can set the latest from 1 to 3 if we receive data from layer 3 before 2.

Reply:
and again, for a given layer, the status may not be set to anything. if there were out of order hare, then we could skip processed layer from N to N+5. same for verified layer.

Reply:
> so if layer 100 is set to latest first, and then set to processed later. when the app asked for latest layer, do you return 99 or 100?

100. because applied > processed > latest

> and again, for a given layer, the status may not be set to anything. if there were out of order hare, then we could skip processed layer from N to N+5. same for verified layer.

is it the case for verified? i thought it is always applied in order
but yes i can see how it can be confusing for latest and processed, i would probably fix that as well
ok i will add one more table